jetstream_consumer.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nats
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+)
+
+func (nc *Conn) createOrUpdateConsumer(opts *jsOpts) (*ConsumerInfo, error) {
+	if opts.streamName == "" {
+		// TODO make err constants
+		return nil, fmt.Errorf("nats: Stream name is required")
+	}
+	if opts.consumer == nil {
+		// TODO make err constants
+		return nil, fmt.Errorf("nats: Consumer configuration is required")
+	}
+
+	crj, err := json.Marshal(&JSApiConsumerCreateRequest{
+		Stream: opts.streamName,
+		Config: *opts.consumer,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := opts.context(nc.Opts.JetStreamTimeout)
+	defer cancel()
+
+	var subj string
+	switch len(opts.consumer.Durable) {
+	case 0:
+		subj = fmt.Sprintf(JSApiConsumerCreateT, opts.streamName)
+	default:
+		subj = fmt.Sprintf(JSApiDurableCreateT, opts.streamName, opts.consumer.Durable)
+	}
+
+	resp, err := nc.RequestWithContext(ctx, subj, crj)
+	if err != nil {
+		return nil, err
+	}
+
+	cresp := &JSApiConsumerCreateResponse{}
+	err = json.Unmarshal(resp.Data, cresp)
+	if err != nil {
+		return nil, err
+	}
+
+	if cresp.IsError() {
+		return nil, cresp.ToError()
+	}
+
+	return cresp.ConsumerInfo, nil
+}
+
+const (
+	JSApiConsumerCreateT = "$JS.API.CONSUMER.CREATE.%s"
+	JSApiDurableCreateT  = "$JS.API.CONSUMER.DURABLE.CREATE.%s.%s"
+)
+
+type ApiError struct {
```

Comment:
Is it an API Error or a RequestError - `ApiError` means that they called it wrong, but likely if there's some other situation it is not an `ApiError`.

Reply:
These are generic errors we use all over in many different messages from the server responses. Only one here but it’s used many places. I kept it the same so it’s a copy paste when we update the server